bigtable-dataflow-parent/bigtable-dataflow-import/src/main/java/com/google/cloud/bigtable/dataflowimport/HBaseExportJob.java

```diff
@@ -0,0 +1,303 @@
+   */
+  static class ShardScans<T> extends DoFn<BoundedSource<T>, KV<String, BoundedSource<T>>> {
+    private static final long DESIRED_SHARD_SIZE = 1024 * 1024 * 1024;
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      final BoundedSource<T> source = c.element();
+
+      final List<? extends BoundedSource<T>> boundedSources = source
+          .splitIntoBundles(DESIRED_SHARD_SIZE, c.getPipelineOptions());
+
+      LOG.info("Split into {} shards", boundedSources.size());
+
+      // randomize order to make sure that scans are distributed across tablets
+      Collections.shuffle(boundedSources);
+
+      // emit each shard with a random uuid, note that uuids are used instead of counters to prevent misleading the
+      // user into thinking that the filenames are ordered in any way
+      for (BoundedSource<T> sourceBundle : boundedSources) {
+        c.output(KV.of(UUID.randomUUID().toString(), sourceBundle));
+      }
+    }
+  }
+
+  /**
+   * Reads each shard and writes its elements into a SequenceFile with compressed blocks.
+   */
+  static class WriteResultsToSeq extends DoFn<KV<String, BoundedSource<Result>>, Void> {
+    private final String basePath;
+    private final Aggregator<Long, Long> itemCounter;
+    private final BatchCounter readDuration;
+    private final BatchCounter writeDuration;
+    private final BatchCounter totalDuration;
+    private final BatchCounter bytesWritten;
+
+    private transient Configuration hadoopConfig;
+
+    WriteResultsToSeq(String basePath) {
+      this.basePath = basePath;
+
+      itemCounter = createAggregator("itemsProcessed", new SumLongFn());
+      readDuration = new BatchCounter(createAggregator("readDuration(secs)", new SumLongFn()), 1_000_000_000);
+      writeDuration = new BatchCounter(createAggregator("writeDuration(secs)", new SumLongFn()), 1_000_000_000);
+      totalDuration = new BatchCounter(createAggregator("totalDuration(secs)", new SumLongFn()), 1_000_000_000);
+      bytesWritten = new BatchCounter(createAggregator("bytesWritten(MB)", new SumLongFn()), 1024 * 1024);
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+    }
+
+    @Override
+    public void startBundle(Context c) throws Exception {
+      hadoopConfig = new Configuration(false);
+
+      // Not using setStrings to avoid loading StringUtils which tries to access HADOOP_HOME
+      hadoopConfig.set("io.serializations", Joiner.on(',').join(
+          WritableSerialization.class.getName(), ResultSerialization.class.getName())
+      );
+      hadoopConfig.set("fs.gs.project.id", c.getPipelineOptions().as(DataflowPipelineOptions.class).getProject());
+      for (Map.Entry<String, String> entry : HBaseImportIO.CONST_FILE_READER_PROPERTIES.entrySet()) {
+        hadoopConfig.set(entry.getKey(), entry.getValue());
+      }
+
+      super.startBundle(c);
+    }
+
+    @Override
```

Comment:
The overall flow of this method is difficult to follow.  I think the primary culprit is the way timing is done.  Let's discuss how we can improve that.